lib/connections/limiter.go

```diff
@@ -99,50 +205,74 @@ func (lim *limiter) String() string {
 
 // limitedReader is a rate limited io.Reader
 type limitedReader struct {
-	reader  io.Reader
-	limiter *limiter
-	isLAN   bool
+	reader   io.Reader
+	limiter  *limiter
+	isLAN    bool
+	remoteID protocol.DeviceID
 }
 
 func (r *limitedReader) Read(buf []byte) (int, error) {
 	n, err := r.reader.Read(buf)
 	if !r.isLAN || r.limiter.limitsLAN.get() {
-		take(r.limiter.read, n)
+
+		// in case rebuildMap was called
+		r.limiter.mu.Lock()
+		deviceLimiter, ok := r.limiter.deviceReadLimiters.Load(r.remoteID)
+		r.limiter.mu.Unlock()
+		if !ok {
```

Comment:
To be honest, if !ok, I'd panic (or do nothing), as it should clearly be there.
Now it's like a bandaid for a problem we can't explain.

I'd do a if deviceLimiter != nil check in take